File: frontend/src/api/paperTrading.ts
import { useEffect } from 'react';
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { apiClient } from './client';
import type {
  BenchmarkCatalogEntry,
  ClosedPositionListResponse,
  PaperTradeListResponse,
  PaperTradeReconcileResult,
  PaperTradingSession,
  PaperTradingSessionKpis,
  PaperTradingSessionListResponse,
  SessionRunListResponse,
  SessionStatus,
  SessionValueHistoryResponse,
} from '../types/api';

/** How often the order-sync poll re-reconciles while a trade is still open. */
const POLL_INTERVAL_MS = 1500;

/** Order statuses that will never change again (mirrors the backend terminal set). */
const TERMINAL_ORDER_STATUSES = new Set(['filled', 'cancelled', 'rejected']);

/** Whether an order status is terminal, so reconciliation can stop polling. */
export function isTerminalOrderStatus(status: string): boolean {
  return TERMINAL_ORDER_STATUSES.has(status);
}

/** Parameters that identify a paper-trading sessions list query. */
export interface SessionsListParams {
  status?: SessionStatus;
  includeArchived?: boolean;
  limit: number;
}

/** Default sessions list parameters (matches the backend defaults). */
export const DEFAULT_SESSIONS_PARAMS: SessionsListParams = { limit: 50 };

/** Typed query keys for paper-trading queries. */
export const paperTradingKeys = {
  all: ['paper-trading'] as const,
  sessions: (params: SessionsListParams) =>
    ['paper-trading', 'sessions', params] as const,
  trades: (sessionId: string) =>
    ['paper-trading', 'session', sessionId, 'trades'] as const,
  runs: (sessionId: string) =>
    ['paper-trading', 'session', sessionId, 'runs'] as const,
  positions: (sessionId: string) =>
    ['paper-trading', 'session', sessionId, 'positions'] as const,
  valueHistory: (sessionId: string) =>
    ['paper-trading', 'session', sessionId, 'value-history'] as const,
  kpis: (sessionId: string) =>
    ['paper-trading', 'session', sessionId, 'kpis'] as const,
  orderSync: (sessionId: string) =>
    ['paper-trading', 'session', sessionId, 'order-sync'] as const,
  benchmarks: () => ['paper-trading', 'benchmarks'] as const,
};

/** Fetch paper-trading sessions (most recently updated first) plus the total. */
export async function listSessions(
  params: SessionsListParams = DEFAULT_SESSIONS_PARAMS,
): Promise<PaperTradingSessionListResponse> {
  const query: Record<string, string | number | boolean> = {
    limit: params.limit,
  };
  if (params.status) query.status = params.status;
  if (params.includeArchived) query.include_archived = true;
  const { data } = await apiClient.get<PaperTradingSessionListResponse>(
    '/api/v1/paper-trading/sessions',
    { params: query },
  );
  return data;
}

/** Archive a stopped session. Rejects with the axios error (404/409). */
export async function archiveSession(
  sessionId: string,
): Promise<PaperTradingSession> {
  const { data } = await apiClient.post<PaperTradingSession>(
    `/api/v1/paper-trading/sessions/${encodeURIComponent(sessionId)}/archive`,
    {},
  );
  return data;
}

/** Restore an archived session. Rejects with the axios error (404). */
export async function unarchiveSession(
  sessionId: string,
): Promise<PaperTradingSession> {
  const { data } = await apiClient.post<PaperTradingSession>(
    `/api/v1/paper-trading/sessions/${encodeURIComponent(sessionId)}/unarchive`,
    {},
  );
  return data;
}

/** Reconcile a session's non-terminal orders; returns counts + refreshed trades. */
export async function reconcileSession(
  sessionId: string,
): Promise<PaperTradeReconcileResult> {
  const { data } = await apiClient.post<PaperTradeReconcileResult>(
    `/api/v1/paper-trading/sessions/${encodeURIComponent(sessionId)}/reconcile`,
    {},
  );
  return data;
}

/** Fetch a session's trades, most recent first. */
export async function listSessionTrades(
  sessionId: string,
  limit = 100,
): Promise<PaperTradeListResponse> {
  const { data } = await apiClient.get<PaperTradeListResponse>(
    `/api/v1/paper-trading/sessions/${encodeURIComponent(sessionId)}/trades`,
    { params: { limit } },
  );
  return data;
}

/** Fetch a session's run history, most recent first. */
export async function listSessionRuns(
  sessionId: string,
  limit = 50,
): Promise<SessionRunListResponse> {
  const { data } = await apiClient.get<SessionRunListResponse>(
    `/api/v1/paper-trading/sessions/${encodeURIComponent(sessionId)}/runs`,
    { params: { limit } },
  );
  return data;
}

/** Fetch a session's closed positions, most recently exited first. */
export async function listSessionPositions(
  sessionId: string,
  limit = 100,
): Promise<ClosedPositionListResponse> {
  const { data } = await apiClient.get<ClosedPositionListResponse>(
    `/api/v1/paper-trading/sessions/${encodeURIComponent(sessionId)}/positions`,
    { params: { limit } },
  );
  return data;
}

/** Fetch a session's daily value snapshots, oldest date first. */
export async function listSessionValueHistory(
  sessionId: string,
): Promise<SessionValueHistoryResponse> {
  const { data } = await apiClient.get<SessionValueHistoryResponse>(
    `/api/v1/paper-trading/sessions/${encodeURIComponent(sessionId)}/value-history`,
  );
  return data;
}

/** Fetch a session's live performance KPIs (value marked to market on load). */
export async function getSessionKpis(
  sessionId: string,
): Promise<PaperTradingSessionKpis> {
  const { data } = await apiClient.get<PaperTradingSessionKpis>(
    `/api/v1/paper-trading/sessions/${encodeURIComponent(sessionId)}/kpis`,
  );
  return data;
}

/** Fetch the fixed benchmark catalog (`[{id, name}]`). */
export async function listBenchmarks(): Promise<BenchmarkCatalogEntry[]> {
  const { data } = await apiClient.get<BenchmarkCatalogEntry[]>(
    '/api/v1/paper-trading/benchmarks',
  );
  return data;
}

/** Switch a session's benchmark; returns the updated session. */
export async function changeSessionBenchmark(
  sessionId: string,
  benchmark: string,
): Promise<PaperTradingSession> {
  const { data } = await apiClient.put<PaperTradingSession>(
    `/api/v1/paper-trading/sessions/${encodeURIComponent(sessionId)}/benchmark`,
    { benchmark },
  );
  return data;
}

/** React Query hook listing paper-trading sessions. */
export function useSessions(
  params: SessionsListParams = DEFAULT_SESSIONS_PARAMS,
) {
  return useQuery<PaperTradingSessionListResponse>({
    queryKey: paperTradingKeys.sessions(params),
    queryFn: () => listSessions(params),
  });
}

/** React Query hook fetching a session's trades. */
export function useSessionTrades(sessionId: string) {
  return useQuery<PaperTradeListResponse>({
    queryKey: paperTradingKeys.trades(sessionId),
    queryFn: () => listSessionTrades(sessionId),
    enabled: sessionId.length > 0,
  });
}

/** React Query hook fetching a session's run history. */
export function useSessionRuns(sessionId: string) {
  return useQuery<SessionRunListResponse>({
    queryKey: paperTradingKeys.runs(sessionId),
    queryFn: () => listSessionRuns(sessionId),
    enabled: sessionId.length > 0,
  });
}

/** React Query hook fetching a session's closed positions. */
export function useSessionPositions(sessionId: string) {
  return useQuery<ClosedPositionListResponse>({
    queryKey: paperTradingKeys.positions(sessionId),
    queryFn: () => listSessionPositions(sessionId),
    enabled: sessionId.length > 0,
  });
}

/** React Query hook fetching a session's daily value history. */
export function useSessionValueHistory(sessionId: string) {
  return useQuery<SessionValueHistoryResponse>({
    queryKey: paperTradingKeys.valueHistory(sessionId),
    queryFn: () => listSessionValueHistory(sessionId),
    enabled: sessionId.length > 0,
  });
}

/** React Query hook fetching a session's live performance KPIs. */
export function useSessionKpis(sessionId: string) {
  return useQuery<PaperTradingSessionKpis>({
    queryKey: paperTradingKeys.kpis(sessionId),
    queryFn: () => getSessionKpis(sessionId),
    enabled: sessionId.length > 0,
  });
}

/** React Query hook fetching the fixed benchmark catalog. */
export function useBenchmarks() {
  return useQuery<BenchmarkCatalogEntry[]>({
    queryKey: paperTradingKeys.benchmarks(),
    queryFn: () => listBenchmarks(),
  });
}

/**
 * Mutation switching a session's benchmark. On success it invalidates the
 * session's own query, its KPIs, and its value-history so the benchmark figures
 * (return, excess return, chart overlay) refresh against the new series.
 */
export function useChangeSessionBenchmark(sessionId: string) {
  const queryClient = useQueryClient();
  return useMutation<PaperTradingSession, unknown, string>({
    mutationFn: (benchmark: string) =>
      changeSessionBenchmark(sessionId, benchmark),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: paperTradingKeys.all });
      queryClient.invalidateQueries({
        queryKey: paperTradingKeys.kpis(sessionId),
      });
      queryClient.invalidateQueries({
        queryKey: paperTradingKeys.valueHistory(sessionId),
      });
    },
  });
}

/**
 * Reconcile a session's orders on mount, then poll every {@link POLL_INTERVAL_MS}
 * until every returned trade reaches a terminal status. Each reconcile that
 * updates at least one trade invalidates the session's trades, positions, and
 * value-history queries so their panels reflect the new fills. Disabled while
 * `sessionId` is empty.
 */
export function useSessionOrderSync(sessionId: string) {
  const queryClient = useQueryClient();
  const query = useQuery<PaperTradeReconcileResult>({
    queryKey: paperTradingKeys.orderSync(sessionId),
    queryFn: () => reconcileSession(sessionId),
    enabled: sessionId.length > 0,
    refetchInterval: (q) => {
      const trades = q.state.data?.trades;
      if (trades === undefined) return POLL_INTERVAL_MS;
      const anyOpen = trades.some(
        (trade) => !isTerminalOrderStatus(trade.order_status),
      );
      return anyOpen ? POLL_INTERVAL_MS : false;
    },
  });

  const reconciled = query.data?.trades_reconciled;
  const updatedAt = query.dataUpdatedAt;
  useEffect(() => {
    if (!reconciled) return;
    queryClient.invalidateQueries({
      queryKey: paperTradingKeys.trades(sessionId),
    });
    queryClient.invalidateQueries({
      queryKey: paperTradingKeys.positions(sessionId),
    });
    queryClient.invalidateQueries({
      queryKey: paperTradingKeys.valueHistory(sessionId),
    });
  }, [reconciled, updatedAt, queryClient, sessionId]);

  return query;
}

/** Mutation archiving a session; invalidates the sessions list on success. */
export function useArchiveSession() {
  const queryClient = useQueryClient();
  return useMutation<PaperTradingSession, unknown, string>({
    mutationFn: (sessionId: string) => archiveSession(sessionId),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: paperTradingKeys.all });
    },
  });
}

/** Mutation unarchiving a session; invalidates the sessions list on success. */
export function useUnarchiveSession() {
  const queryClient = useQueryClient();
  return useMutation<PaperTradingSession, unknown, string>({
    mutationFn: (sessionId: string) => unarchiveSession(sessionId),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: paperTradingKeys.all });
    },
  });
}

File: backend/src/cadence/api/routers/paper_trading.py
"""Paper-trading resource router. Mounted under ``/api/v1``.

Trade/run/position writes happen via the ai_portfolio flow; this router exposes
the session reads (listing plus a session's trades, run history, and closed
positions) and the session archive/unarchive actions.
"""

from __future__ import annotations

import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from cadence.api.schemas import (
    BenchmarkCatalogEntry,
    ClosedPositionListResponse,
    ClosedPositionRead,
    PaperTradeListResponse,
    PaperTradeRead,
    PaperTradeReconcileRead,
    PaperTradingSessionKpisRead,
    PaperTradingSessionListResponse,
    PaperTradingSessionRead,
    SessionBenchmarkChangeRequest,
    SessionRunListResponse,
    SessionRunRead,
    SessionValueHistoryResponse,
    SessionValueSnapshotRead,
)
from cadence.broker import get_broker
from cadence.broker.base import Broker
from cadence.database import get_db
from cadence.paper_trading import service
from cadence.paper_trading.constants import (
    BENCHMARK_DISPLAY_NAMES,
    Benchmark,
    SessionStatus,
)
from cadence.paper_trading.errors import (
    InvalidBenchmarkError,
    SessionNotArchivableError,
    SessionNotFoundError,
)

router = APIRouter(prefix="/paper-trading", tags=["paper-trading"])

DbSession = Annotated[Session, Depends(get_db)]
BrokerDep = Annotated[Broker, Depends(get_broker)]


@router.get("/sessions", response_model=PaperTradingSessionListResponse)
def list_sessions(
    db: DbSession,
    status_filter: Annotated[
        SessionStatus | None,
        Query(alias="status", description="Filter by session status"),
    ] = None,
    include_archived: Annotated[
        bool,
        Query(description="Include archived sessions in the result"),
    ] = False,
    limit: int = 50,
) -> PaperTradingSessionListResponse:
    """Return paper-trading sessions, most recently updated first."""
    items = [
        PaperTradingSessionRead.model_validate(row)
        for row in service.list_sessions(
            db,
            status=status_filter,
            include_archived=include_archived,
            limit=limit,
        )
    ]
    total = service.count_sessions(
        db, status=status_filter, include_archived=include_archived
    )
    return PaperTradingSessionListResponse(items=items, total=total)


def _require_session(db: Session, session_id: uuid.UUID) -> None:
    """Raise 404 if the session does not exist."""
    try:
        service.get_session(db, session_id)
    except SessionNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)
        ) from exc


@router.post(
    "/sessions/{session_id}/archive", response_model=PaperTradingSessionRead
)
def archive_session(
    session_id: uuid.UUID, db: DbSession
) -> PaperTradingSessionRead:
    """Soft-archive a stopped session (409 if it is not stopped)."""
    try:
        row = service.archive_session(db, session_id)
    except SessionNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)
        ) from exc
    except SessionNotArchivableError as exc:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT, detail=str(exc)
        ) from exc
    return PaperTradingSessionRead.model_validate(row)


@router.post(
    "/sessions/{session_id}/unarchive", response_model=PaperTradingSessionRead
)
def unarchive_session(
    session_id: uuid.UUID, db: DbSession
) -> PaperTradingSessionRead:
    """Restore an archived session to the default listing."""
    try:
        row = service.unarchive_session(db, session_id)
    except SessionNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)
        ) from exc
    return PaperTradingSessionRead.model_validate(row)


@router.post(
    "/sessions/{session_id}/reconcile", response_model=PaperTradeReconcileRead
)
def reconcile_session(
    session_id: uuid.UUID, db: DbSession, broker: BrokerDep
) -> PaperTradeReconcileRead:
    """Reconcile a session's non-terminal orders against the broker (404 if unknown).

    Returns the reconciliation counts plus the session's refreshed trades so the
    client can render the updated statuses in one round-trip.
    """
    try:
        result = service.reconcile_session_orders(db, broker, session_id)
    except SessionNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)
        ) from exc
    trades = [
        PaperTradeRead.model_validate(row)
        for row in service.get_session_trades(db, session_id)
    ]
    return PaperTradeReconcileRead(
        trades_seen=result.trades_seen,
        trades_reconciled=result.trades_reconciled,
        trades_filled=result.trades_filled,
        trades_basis_corrected=result.trades_basis_corrected,
        trades=trades,
    )


@router.get(
    "/sessions/{session_id}/trades", response_model=PaperTradeListResponse
)
def list_session_trades(
    session_id: uuid.UUID,
    db: DbSession,
    limit: int = 100,
) -> PaperTradeListResponse:
    """Return a session's trades, most recent first."""
    _require_session(db, session_id)
    items = [
        PaperTradeRead.model_validate(row)
        for row in service.get_session_trades(db, session_id, limit=limit)
    ]
    total = service.count_session_trades(db, session_id)
    return PaperTradeListResponse(items=items, total=total)


@router.get("/sessions/{session_id}/runs", response_model=SessionRunListResponse)
def list_session_runs(
    session_id: uuid.UUID,
    db: DbSession,
    limit: int = 50,
) -> SessionRunListResponse:
    """Return a session's run history, most recent first."""
    _require_session(db, session_id)
    items = [
        SessionRunRead.model_validate(row)
        for row in service.get_session_runs(db, session_id, limit=limit)
    ]
    total = service.count_session_runs(db, session_id)
    return SessionRunListResponse(items=items, total=total)


@router.get(
    "/sessions/{session_id}/positions",
    response_model=ClosedPositionListResponse,
)
def list_session_positions(
    session_id: uuid.UUID,
    db: DbSession,
    limit: int = 100,
) -> ClosedPositionListResponse:
    """Return a session's closed positions, most recently exited first."""
    _require_session(db, session_id)
    items = [
        ClosedPositionRead.model_validate(row)
        for row in service.get_closed_positions(db, session_id, limit=limit)
    ]
    total = service.count_closed_positions(db, session_id)
    return ClosedPositionListResponse(items=items, total=total)


@router.get(
    "/sessions/{session_id}/kpis",
    response_model=PaperTradingSessionKpisRead,
)
def get_session_kpis(
    session_id: uuid.UUID, db: DbSession, broker: BrokerDep
) -> PaperTradingSessionKpisRead:
    """Return a session's live performance KPIs (404 if unknown).

    Marks the session's open positions to market via the broker on each call, so
    the current value and unrealised P&L reflect current quotes rather than the
    last stored snapshot.
    """
    try:
        kpis = service.session_kpis(db, session_id=session_id, broker=broker)
    except SessionNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)
        ) from exc
    return PaperTradingSessionKpisRead(
        current_value=kpis.current_value,
        realised_pnl=kpis.realised_pnl,
        unrealised_pnl=kpis.unrealised_pnl,
        total_fees=kpis.total_fees,
        total_return=kpis.total_return,
        total_return_pct=kpis.total_return_pct,
        sharpe_ratio=kpis.sharpe_ratio,
        benchmark=kpis.benchmark,
        benchmark_return_pct=kpis.benchmark_return_pct,
        excess_return_pct=kpis.excess_return_pct,
        excess_return=kpis.excess_return,
    )


@router.get(
    "/sessions/{session_id}/value-history",
    response_model=SessionValueHistoryResponse,
)
def list_session_value_history(
    session_id: uuid.UUID,
    db: DbSession,
) -> SessionValueHistoryResponse:
    """Return a session's daily value snapshots (with benchmark), oldest first."""
    _require_session(db, session_id)
    items = [
        SessionValueSnapshotRead.model_validate(point.snapshot).model_copy(
            update={"benchmark_value": point.benchmark_value}
        )
        for point in service.list_value_history(db, session_id=session_id)
    ]
    return SessionValueHistoryResponse(items=items, total=len(items))


@router.get("/benchmarks", response_model=list[BenchmarkCatalogEntry])
def list_benchmarks() -> list[BenchmarkCatalogEntry]:
    """Return the fixed benchmark catalog as ``[{id, name}]``."""
    return [
        BenchmarkCatalogEntry(id=member.value, name=BENCHMARK_DISPLAY_NAMES[member])
        for member in Benchmark
    ]


@router.put(
    "/sessions/{session_id}/benchmark",
    response_model=PaperTradingSessionRead,
)
def change_session_benchmark(
    session_id: uuid.UUID,
    payload: SessionBenchmarkChangeRequest,
    db: DbSession,
) -> PaperTradingSessionRead:
    """Switch a session's benchmark (404 unknown session, 422 invalid id)."""
    try:
        row = service.change_session_benchmark(
            db, session_id=session_id, benchmark=payload.benchmark
        )
    except SessionNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)
        ) from exc
    except InvalidBenchmarkError as exc:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc)
        ) from exc
    return PaperTradingSessionRead.model_validate(row)
